### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by mattia-p**
**Firmware**
Changed the delay at the end of the loop from 100ms to 5ms, to reduce the delay between the send the commands to the actuators.

**Target generator**
1. Added the possibility to loop over waypoints (go from the last one to the first one). 
2. Improved the way we handle the last waypoint. When the vehicle gets at the last waypoint, the target generator used to set the target waypoint as one of the waypoints behind, which causes the vehicle to steer towards it. Instead we detect that the vehicle is at the last waypoint and we freeze the steering to zero. 

**RTK GPS serial reader**
Added error handling. If parsing the nmea sentence doesn't work, we don't publish anything and try again with the next message, instead of breaking the node.

**Lateral controller**
Added timed loop that runs most of the node, instead of running everything from the callback.

### CODE REVIEW

File: autonomy_software/controllers_pkg/src/controllers_pkg/lateral_controller.py [hunk 1]
```diff
@@ -80,38 +83,51 @@ def current_pose_callback(self, msg: PoseStamped):
...
+            self.rate.sleep()
 
 
 if __name__ == "__main__":
     try:
         rospy.init_node("lateral_controller")
         lateral_controller = LateralController()
-        rospy.spin()
+        # rospy.spin()
+        lateral_controller.loop()
```
Comment (roux-antoine) [inline autonomy_software/controllers_pkg/src/controllers_pkg/lateral_controller.py:131]: Will this cause us to publish almost continuously? If no, maybe we want to implement some sort of rate limitation just to make sure we don't spam the Arduino too much.

File: autonomy_software/localization_pkg/src/rtk_gps_serial_reader.py [hunk 2]
```diff
@@ -39,9 +39,17 @@ def loop(self):
...
+                    try:
+                        parsed_msg = NMEAReader.parse(serial_data)
+                        print(parsed_msg)
+                        self.publish_nmea_sentence_rmc(parsed_msg)
+                    except Exception:
+                        rospy.logwarn(
+                            "Failed to parse NMEA sentence. Input serial data: {}. Not publishing /gps_info.".format(
+                                serial_data
+                            )
+                        )
```
Comment (roux-antoine) [inline autonomy_software/localization_pkg/src/rtk_gps_serial_reader.py:52]:
```suggestion
                        rospy.logerror(f"Failed to parse NMEA sentence. Input serial data: {serial_data}. Not publishing /gps_info.")
```

File: autonomy_software/planning_pkg/src/target_generator.py [hunk 3]
```diff
@@ -127,17 +129,38 @@ def callback_current_pose(self, pose_msg: PoseStamped):
...
             nextWaypointId = self.getNextWaypoint()
 
-            """ Get target point """
-            targetPoint = self.getTargetPoint(nextWaypointId)
+            # If next waypoint is the last one and we are not looping over, no need to
+            # try and find the target point. If we do, the target point will be behind us
+            # and this is a potential hazard. Instead, set target point to last waypoint.
+            # NOTE: This could theoretically be better if getNextWaypoint would publish something else when
+            # all the waypoints are within the lookahead
+            if nextWaypointId is None and self.loop_over_waypoints is False:
```
Comment (roux-antoine) [inline autonomy_software/planning_pkg/src/target_generator.py:137]: Do you have the githooks enabled? I think black would suggest modifications here (maybe something like `if not nextWaypointId and not self.loop_over_waypoints:`). Not a big issue.

File: autonomy_software/planning_pkg/src/target_generator.py [hunk 4]
```diff
@@ -212,11 +238,24 @@ def getNextWaypoint(self):
...
+                    plane_distance(
+                        State(x=wp.pose.position.x, y=wp.pose.position.y),
+                        self.current_state,
+                    )
+                    > self.lookahead_distance
+                ):
+                    return wp.id
+
+        # If we are here, all the waypoints are within the lookahead distance
+        # In this case, we return the last waypoint
```
Comment (roux-antoine) [inline autonomy_software/planning_pkg/src/target_generator.py:256]: This comment can be updated now that we don't return the last waypoint

File: autonomy_software/planning_pkg/src/target_generator.py (hunk 3 above)
Comment (mattia-p) [reply autonomy_software/planning_pkg/src/target_generator.py:137]: I do yes. It actually suggested me that I think. I used to have something different. Should we change?

File: autonomy_software/controllers_pkg/src/controllers_pkg/lateral_controller.py (hunk 1 above)
Comment (mattia-p) [reply autonomy_software/controllers_pkg/src/controllers_pkg/lateral_controller.py:131]: Correct. This makes us publish the steering command at 10Hz.
Do you mean "If YES, maybe implement some sort of rate limitation...'?

File: autonomy_software/planning_pkg/src/target_generator.py (hunk 4 above)
Comment (mattia-p) [reply autonomy_software/planning_pkg/src/target_generator.py]: Done

File: autonomy_software/localization_pkg/src/rtk_gps_serial_reader.py (hunk 2 above)
Comment (mattia-p) [reply autonomy_software/localization_pkg/src/rtk_gps_serial_reader.py]: Done

File: autonomy_software/controllers_pkg/src/controllers_pkg/lateral_controller.py (hunk 1 above)
Comment (roux-antoine) [reply autonomy_software/controllers_pkg/src/controllers_pkg/lateral_controller.py:131]: My bad I had misunderstood the code, lgtm

File: autonomy_software/planning_pkg/src/target_generator.py (hunk 3 above)
Comment (roux-antoine) [reply autonomy_software/planning_pkg/src/target_generator.py:137]: No that's fine

**Review by roux-antoine — APPROVED** [review]

**Pull request merged**